### PR TITLE
Feature/matrix inverse autodiff

### PR DESCRIPTION
**Pull request by christophernorth**
More Testing (in terms of finite differences etc) will come in as I add for others at the same time.

### CODE REVIEW

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixInverseVertexTest.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+            -1.5, 0.75,
+            0.5, -0.25},
+            new int[]{2, 2, 2, 2}
+        );
+
+        assertEquals(expectedInverseWrtMatrix, inverseWrtMatrix);
+    }
+
+    @Test
+    public void inverseMultipliedEqualsIdentity() {
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixInverseVertexTest.java]: Nice test!

File: keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixInverseVertexTest.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+import io.improbable.keanu.tensor.dbl.ScalarDoubleTensor;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.UniformVertex;
+
+public class MatrixInverseVertexTest {
+
+    @Test(expected = IllegalArgumentException.class)
+    public void rejectsNonSquareInput() {
+        DoubleTensor matrix = DoubleTensor.arange(1, 9).reshape(4, 2);
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixInverseVertexTest.java]: Nice exception checking here 👍

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java [hunk 3]
```diff
@@ -169,6 +169,20 @@ public DualNumber matrixMultiplyBy(DualNumber that) {
...
     }
 
+    public DualNumber matrixInverse() {
+        //dc = -A^-1 * da * A^-1
+        DoubleTensor newValue = this.value.matrixInverse();
+
+        if (this.partialDerivatives.isEmpty()) {
+            return new DualNumber(newValue, PartialDerivatives.OF_CONSTANT);
+        } else {
+            DoubleTensor negatedValue = newValue.unaryMinus();
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: Nicely done with the matrix multiply.

File: keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/ScalarDoubleTensor.java [hunk 4]
```diff
@@ -329,8 +329,8 @@ public DoubleTensor max(DoubleTensor max) {
     }
 
     @Override
-    public DoubleTensor inverse() {
-        return duplicate();
+    public DoubleTensor matrixInverse() {
+        return new ScalarDoubleTensor(1 / value);
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/ScalarDoubleTensor.java]: Good find! We need some better scalar testing...

Comment (christophernorth) [reply keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/ScalarDoubleTensor.java]: Yep - I intend to add some in as part of my testing - it's definitely our weakest area by far at the moment.  Potentially a crazy idea - but should we disable the use of the Scalar objects until we have better testing there?

**Comment by christophernorth** [comment]
Fixes #231

**Pull request merged**